SSA/Projects/ModArith/Basic.lean

```diff
@@ -0,0 +1,150 @@
+/-
+This file contains the definition of the MLIR `ModArith` dialect as
+implemented in HEIR, see:
+  https://heir.dev/docs/dialects/modarith/
+
+It is structurally similar to `FullyHomomorphicEncryption.Basic.lean` but focuses on arithmetic
+directly in ℤ/qℤ (ZMod q), rather than polynomials over ℤ/qℤ.
+
+Authors: Jaeho Choi<zerozerozero0216@gmail.com>
+-/
+import Mathlib.Data.ZMod.Basic
+import SSA.Core.Framework
+
+open ZMod
+
+/-!
+  # ModArith Dialect
+
+  The `ModArith` dialect is a simpler variant that models integer arithmetic
+  **modulo `q`**, i.e., arithmetic in the ring `ZMod q`.
+
+  We assume `q > 1` as a fact. We denote our base ring as: `R = ZMod q`.
+
+  The dialect's type system includes (for example) `index`, `integer`, and
+  a specialized type `modLike` for elements in `ZMod q`.
+
+  Operations: Add, Sub, Mul, and constants for this ring, plus integers and
+  indices.
+-/
+
+section CommRing
+/- We assume `q > 1`. -/
+variable (q : ℕ) [Fact (q > 1)]
+
+/--
+By analogy to `R q n` from the `Poly` dialect, we simply define
+`R q := ZMod q`. The ring structure on `ZMod q` is already known
+to mathlib.
+-/
+abbrev R := ZMod q
+
+end CommRing
+
+/-!
+## Dialect type definitions
+
+Here, we define a small type system for the `ModArith` dialect:
+  1. `index` – if you also want natural-number indices (mirroring MLIR’s index).
+  2. `integer` – for full-range integers in Lean (ℤ).
+  3. `modLike` – for our ring `ZMod q`.
+
+You can freely add more types or rename them according to your needs.
+-/
+inductive Ty (q : ℕ) where
+| index
+| integer
+| modLike
+deriving DecidableEq, Repr
+
+instance : Inhabited (Ty q) := ⟨Ty.index⟩
```

Comment:
Assuming the specific default type doesn't matter, you could also derive `Inhabited`

```suggestion
deriving DecidableEq, Repr, Inhabited
```